pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/ServerRequestPlanVisitor.java

```diff
@@ -0,0 +1,363 @@
+    final DataSchema.ColumnDataType columnDataType = block.getDataSchema().getColumnDataType(idx);
+    final FieldSpec.DataType storedType = columnDataType.getStoredType().toDataType();;
+    IdSet idSet = IdSets.create(columnDataType.toDataType());
+    for (Object[] row : block.getContainer()) {
+      switch (storedType) {
+        case INT:
+          idSet.add((int) row[idx]);
+          break;
+        case LONG:
+          idSet.add((long) row[idx]);
+          break;
+        case FLOAT:
+          idSet.add((float) row[idx]);
+          break;
+        case DOUBLE:
+          idSet.add((double) row[idx]);
+          break;
+        case STRING:
+          idSet.add((String) row[idx]);
+          break;
+        default:
+          throw new IllegalStateException("Illegal SV data type for ID_SET aggregation function: " + storedType);
+      }
+    }
+    try {
+      return RequestUtils.getLiteralExpression(idSet.toBase64String());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Helper method to attach the time boundary to the given PinotQuery.
+   */
+  private static void attachTimeBoundary(PinotQuery pinotQuery, TimeBoundaryInfo timeBoundaryInfo,
+      boolean isOfflineRequest) {
+    String timeColumn = timeBoundaryInfo.getTimeColumn();
+    String timeValue = timeBoundaryInfo.getTimeValue();
+    Expression timeFilterExpression = RequestUtils.getFunctionExpression(
+        isOfflineRequest ? FilterKind.LESS_THAN_OR_EQUAL.name() : FilterKind.GREATER_THAN.name());
+    timeFilterExpression.getFunctionCall().setOperands(
+        Arrays.asList(RequestUtils.getIdentifierExpression(timeColumn), RequestUtils.getLiteralExpression(timeValue)));
+    attachFilterExpression(pinotQuery, FilterKind.AND, Collections.singletonList(timeFilterExpression));
+  }
+
+  private static void attachFilterExpression(PinotQuery pinotQuery, FilterKind attachKind, List<Expression> exprs) {
+    Preconditions.checkState(attachKind == FilterKind.AND || attachKind == FilterKind.OR);
+    Expression filterExpression = pinotQuery.getFilterExpression();
+    List<Expression> arrayList = new ArrayList<>(exprs);
+    if (filterExpression != null) {
+      arrayList.add(filterExpression);
+    }
+    if (arrayList.size() > 1) {
+      Expression attachFilterExpression = RequestUtils.getFunctionExpression(attachKind.name());
+      attachFilterExpression.getFunctionCall().setOperands(arrayList);
+      pinotQuery.setFilterExpression(attachFilterExpression);
+    } else {
+      pinotQuery.setFilterExpression(arrayList.get(0));
+    }
+  }
+
+  private static TransferableBlock attachDynamicOperatorResults(BaseOperator<TransferableBlock> baseOperator,
+      DataSchema dataSchema) {
+    TransferableBlock mergedBlock = new TransferableBlock(new ArrayList<>(), dataSchema, BaseDataBlock.Type.ROW);
+    TransferableBlock block = baseOperator.nextBlock();
```

Comment:
there should be a way although i knew we have to touch a lot of v1's code.
- getting the data block afterwards means we put a placeholder / future reference, not sure (1) where to put it (maybe QueryContext), and (2) whether transform/filter operators allows those (only literal/expression are supported i think, we need another type)
- i don't know if in the future we want to 
    - use any mailbox-received content elsewhere in other operators
    - use any mailbox-received contents in streaming fashion (e.g. don't wait for the full block to complete)

but regardless, a design doc is needed if we were to resolve this later. now I have no idea

Reply:
what if we built a physical operator for DynamicFilter? it could be a filter that reads from two mailboxes similar to the join operator which first reads from the broadcast table and then reads from the other side.

Reply:
the whole point for dynamic filter is to leverage the filter code from leaf stage that benefit from index. 
e.g. we don't "read" from the "otherside", we apply index directly (along with all other filters) to get a set of rowIDs, then we bring those selected data into heap. 

so regardless of how we solve, (1) we probably have to model it as a new physical operator call dynamicfilter anyway, but (2) it needs to be applied together with all other filters. 

if we bring in the rows on heap then apply a generic filter function, it might not be so much faster than the normal inner join process.